pkg/v1/tarball/layer.go

```diff
@@ -90,11 +102,26 @@ func (l *layer) MediaType() (types.MediaType, error) {
 // LayerOption applies options to layer
 type LayerOption func(*layer)
 
+// WithCompression is a functional option for overriding the default
+// compression algorithm used for compressing uncompressed tarballs.
+// Also updates the mediaType to "application/vnd.oci.image.layer.v1.tar+zstd"
+// if zstd compression is selected.
+func WithCompression(compression LayerCompression) LayerOption {
+	return func(l *layer) {
+		if compression == ZStd {
+			l.mediaType = types.OCILayerZStd
+		} else {
+			l.mediaType = types.DockerLayer
```

Comment:
I would expect `WithCompression(None)` to set `l.mediaType` to `DockerUncompressedLayer` and produce uncompressed layers. WDYT?

I am also OK with punting on this for now, returning an error or documenting that None is invalid.

Reply:
That makes sense, but I'm not quite sure how easy this would be to implement. My understanding is that the decision whether layers are produced with or without compression is made by consumers of the Layer interface when they access the data via `.Compressed()` or `.Uncompressed()`. Hence I'd have to modify all the code sites that use `.Compressed()` to make `WithCompression(None)` work, right?

I was thinking about alternatively just returning uncompressed data from the `.Compressed()` method when `WithCompression(None)` is set. But I felt that would conflict with semantics of the `.Compressed()` method which may be a bit confusing. Then again, this PR assumes the consumers of the `.Compressed()` method are agnostic to the compression format anyway, so one could argue that no compression can still be regarded "some kind of very inefficient compression format".

If you agree with this idea, I would implement it that way. Otherwise, I think the best option might be to drop `WithCompression(None)` altogether for now.

Reply:
> But I felt that would conflict with semantics of the .Compressed() method which may be a bit confusing.

Yeah that's fair. Having both the `Compressed` and `Uncompressed` methods is a design mistake. It would be better if we deprecated `Uncompressed` and renamed `Compressed`, but it's too late for that. I am fine with dropping this for now, but I think we want to return an error if we don't want to support it.

Reply:
I modified the `WithCompression` method to ignore calls the `compression.None` argument and use `gzip` instead. Additionally it will print this warning:
```
Compression type 'none' is not supported for tarball layers; using gzip compression.
```